benchmark/preconditioner/preconditioner.cpp

```diff
@@ -159,45 +159,47 @@ void run_preconditioner(const char *precond_name,
                               allocator);
         }
 
+        auto generate_timer = get_timer(exec, FLAGS_gpu_timer);
+        auto apply_timer = get_timer(exec, FLAGS_gpu_timer);
+
         {
             // fast run, gets total time
             auto x_clone = clone(x);
 
             auto precond = precond_factory.at(precond_name)(exec);
 
-            for (auto i = 0u; i < FLAGS_warmup; ++i) {
+            IterationControl ic_gen{generate_timer};
+            IterationControl ic_apply{apply_timer};
+
+            for (auto _ : ic_apply.warmup_run()) {
                 precond->generate(system_matrix)->apply(lend(b), lend(x_clone));
             }
-            auto generate_timer = get_timer(exec, FLAGS_gpu_timer);
-            auto apply_timer = get_timer(exec, FLAGS_gpu_timer);
 
-            exec->synchronize();
-            generate_timer->tic();
             std::unique_ptr<gko::LinOp> precond_op;
-            for (auto i = 0u; i < FLAGS_repetitions; ++i) {
+            for (auto _ : ic_gen.run()) {
+                exec->synchronize();
+                generate_timer->tic();
                 precond_op = precond->generate(system_matrix);
+                generate_timer->toc();
             }
-            generate_timer->toc();
```

Comment:
In Nonius/Catch2, they measure the overhead of the timer in a separate loop, we could do the same thing here?

Reply:
The problem is that is very dynamic, what changes is one synchronization after many kernel launches vs one synchronization per kernel for the GPU timing. I don't think we can hope to accurately estimate that sadly. Maybe we could change the timer to optionally take a number of iterations information from the outside instead of relying on `tic` and `toc` inside the loop?

Reply:
Kind of a two-stage approach, only measuring every n-th time?

Reply:
That would be one way to do it I think, maybe there are other ways as well.

Note that technically, that could also happen for some SpMV and BLAS depending on the parameters I guess. For solvers I think we would not see too much a difference because there is already one synchronization with a few byte being sent in every iteration of the solver (the booleans for the stopping condition).

Reply:
I would go for an approach similar to the first version of this PR, i.e. increasing the number of iterations between two timings until the benchmark is finished. That would result in fewer timings, but still more than one.

Also, that would mean that some part in `IterationControl` can issue timings. Now the question is, should there still be timings outside `IterationControl` or not? To clarify, I think the version with timings outside would look like
```c++
timer->tic();
for(auto _: ic.run()){
  op->apply(...);
}
timer->toc();
auto reps = ic.get_repetitions(); // or something similar
```
or alternatively
```c++
for(auto _: ic.run()){
  op->apply(...);
}
auto runtime = ic.get_average_time();
auto reps = ic.get_repetitions();
```
I think the version without outside timings would be easier to handle, but that's just a guess.

Reply:
I've implemented a version that increases the number of iterations between two timings. For very small examples, this reduces the number of timing call quite a lot, e.g. BLAS benchmark with `n=1000` uses ~18000 repetitions but only 24 timing calls.

Disadvantage: Everything within the `for(auto _: ic.run())` loop is timed. In the solver benchmark this leads to additionally taking the timing of `x_clone = clone(x)`. IMO that should not disturb the benchmark too much.

Reply:
Currently, there is no `exec->synchronize()` call before `timer->tic()`. The CPUTimer automatically synchronizes in its `tic_impl`, should the CUDA, HIP timers also do this?

Reply:
It's interesting that the CPU does this `exec->synchronize()` on `tic`. Anyway that case is mostly for cosmetics because the CPU synchronization should do nothing. I think you can do it for the GPU as well it shouldn't hurt, but it shouldn't change anything either since there shouldn't be any kernel non-synchronized already I think?

I think the solvers can be left like the previous version without timing the `clone(x)` if possible. There shouldn't be a difference in terms of timing overhead due to already doing synchronizations inside the solver apply. The `clone` I believe could be quite impactful since that depends on how reactive the CUDA runtime library is, which in some cases can have weird behaviors.

Reply:
I think it is not possible to reduce the number of timings, if the `clone(x)` should be left out of the timings. That said, the timing overhead didn't matter for the solver benchmark anyways. I will try to figure out how to keep both behaviors (timing in each iteration vs variable intervals between timings).

Reply:
Yes certainly, for solvers I don't expect to win anything here, we will need 1 timing per loop iteration, but as said I think that should be fine. The big issue as you say is we need a way to have both behavior coexisting, hopefully with a not too confusing interface/code.

Reply:
I've added a switch for the behavior change: 
- `ic.run(true)` will manage the timing itself, i.e. everything within the loop will be timed
- `ic.run(false)` does not manage the timing, i.e. the timing has to be issued by the user.

The default case `ic.run()` will use `ic.run(true)`. I think the interface/documentation should be clear enough. (However, the implementation became a bit more messy).